PayloadPkg/Library/PayloadEntryLib/PayloadEntryLib.c

```diff
@@ -78,20 +80,32 @@ PayloadInit (
   GetPayloadReservedRamRegion (&RsvdBase, &RsvdSize);
   ASSERT ((RsvdBase & EFI_PAGE_MASK) == 0);
 
+  if (FeaturePcdGet (PcdDmaProtectionEnabled)) {
+    DmaSize = ALIGN_UP (PcdGet32 (PcdDmaBufferSize), EFI_PAGE_SIZE);
+  } else {
+    DmaSize = 0;
+  }
+  DmaBase  = (UINT32)RsvdBase - DmaSize;
+  DmaBase  = ALIGN_DOWN(DmaBase, PcdGet32 (PcdDmaBufferAlignment));
+
   HeapSize  = ALIGN_UP (PcdGet32 (PcdPayloadHeapSize), EFI_PAGE_SIZE);
-  HeapBase = (UINT32)RsvdBase - HeapSize;
+  HeapBase = DmaBase - HeapSize;
 
   StackSize = ALIGN_UP (PcdGet32 (PcdPayloadStackSize), EFI_PAGE_SIZE);
   StackBase = HeapBase - StackSize;
 
   // Add payload reserved memory region and free memory region
+  // Use EfiRuntimeServicesData as DMA memory pool
   MemoryRanges[0].BaseAddress   = HeapBase;
   MemoryRanges[0].NumberOfPages = EFI_SIZE_TO_PAGES (HeapSize);
   MemoryRanges[0].Type          = EfiBootServicesData;
   MemoryRanges[1].BaseAddress   = RsvdBase;
   MemoryRanges[1].NumberOfPages = EFI_SIZE_TO_PAGES ((UINT32)RsvdSize);
   MemoryRanges[1].Type          = EfiReservedMemoryType;
-  AddMemoryResourceRange (MemoryRanges, 2);
+  MemoryRanges[2].BaseAddress   = DmaBase;
+  MemoryRanges[2].NumberOfPages = EFI_SIZE_TO_PAGES (DmaSize);
+  MemoryRanges[2].Type          = EfiRuntimeServicesData;
+  AddMemoryResourceRange (MemoryRanges, 3);

```

Comment:
if SplitMemroyMap() is updated, I think we don't need update it here in the payload.

Reply:
This is still required.  It needs to be added into memory pool as a different memory type.